# Native-only precise HBM sound includes, modern pure destructor syntax, and pointer-rounding ABI.
# Upstream-base: 7a6bf165d9f1c5914181c079c7d4bfa7204f3509
# Prerequisite source523; composed compiler proof also uses frozen524/527/528/530/531/533.

diff --git a/libs/RVL_SDK/include/revolution/hbm/nw4hbm/snd/BankFile.h b/libs/RVL_SDK/include/revolution/hbm/nw4hbm/snd/BankFile.h
index 3d469f8..24108a1 100644
--- a/libs/RVL_SDK/include/revolution/hbm/nw4hbm/snd/BankFile.h
+++ b/libs/RVL_SDK/include/revolution/hbm/nw4hbm/snd/BankFile.h
@@ -6,7 +6,11 @@
 #include "revolution/hbm/nw4hbm/snd/Util.h"
 #include "revolution/hbm/nw4hbm/snd/WaveFile.h"
 
+#if defined(MSCHARGED_NATIVE)
+#include "revolution/hbm/nw4hbm/ut/binaryFileFormat.h"
+#else
 #include "revolution/hbm/nw4hbm/ut.h"
+#endif
 
 #include "revolution/hbm/HBMCommon.h"
 
diff --git a/libs/RVL_SDK/include/revolution/hbm/nw4hbm/snd/SeqPlayer.h b/libs/RVL_SDK/include/revolution/hbm/nw4hbm/snd/SeqPlayer.h
index 8f594c5..f08adbb 100644
--- a/libs/RVL_SDK/include/revolution/hbm/nw4hbm/snd/SeqPlayer.h
+++ b/libs/RVL_SDK/include/revolution/hbm/nw4hbm/snd/SeqPlayer.h
@@ -2,6 +2,9 @@
 #define NW4HBM_SND_SEQ_PLAYER_H
 
 #include "revolution/types.h"
+#if defined(MSCHARGED_NATIVE)
+#include "revolution/hbm/nw4hbm/ut/Lock.h"
+#endif
 
 #include "revolution/hbm/nw4hbm/snd/BasicPlayer.h"
 
diff --git a/libs/RVL_SDK/include/revolution/hbm/nw4hbm/snd/SeqSoundHandle.h b/libs/RVL_SDK/include/revolution/hbm/nw4hbm/snd/SeqSoundHandle.h
index 4ef8942..988c76d 100644
--- a/libs/RVL_SDK/include/revolution/hbm/nw4hbm/snd/SeqSoundHandle.h
+++ b/libs/RVL_SDK/include/revolution/hbm/nw4hbm/snd/SeqSoundHandle.h
@@ -7,7 +7,11 @@
 #include "revolution/hbm/nw4hbm/snd/SeqPlayer.h"
 #include "revolution/hbm/nw4hbm/snd/SeqSound.h"
 
+#if defined(MSCHARGED_NATIVE)
+#include "revolution/hbm/nw4hbm/ut/inlines.h"
+#else
 #include "revolution/hbm/nw4hbm/ut.h"
+#endif
 
 namespace nw4hbm {
 namespace snd {
diff --git a/libs/RVL_SDK/include/revolution/hbm/nw4hbm/snd/SoundArchiveLoader.h b/libs/RVL_SDK/include/revolution/hbm/nw4hbm/snd/SoundArchiveLoader.h
index ac1c50d..44b0b80 100644
--- a/libs/RVL_SDK/include/revolution/hbm/nw4hbm/snd/SoundArchiveLoader.h
+++ b/libs/RVL_SDK/include/revolution/hbm/nw4hbm/snd/SoundArchiveLoader.h
@@ -2,7 +2,12 @@
 #define NW4HBM_SND_SOUND_ARCHIVE_LOADER_H
 
 #include "revolution/hbm/nw4hbm/snd/types.h"
+#if defined(MSCHARGED_NATIVE)
+#include "revolution/hbm/nw4hbm/ut/FileStream.h"
+#include "revolution/hbm/HBMAssert.hpp"
+#else
 #include "revolution/hbm/nw4hbm/ut.h"
+#endif
 #include "revolution/os.h"
 
 namespace nw4hbm {
diff --git a/libs/RVL_SDK/include/revolution/hbm/nw4hbm/ut/FileStream.h b/libs/RVL_SDK/include/revolution/hbm/nw4hbm/ut/FileStream.h
index 5b32402..578bfd0 100644
--- a/libs/RVL_SDK/include/revolution/hbm/nw4hbm/ut/FileStream.h
+++ b/libs/RVL_SDK/include/revolution/hbm/nw4hbm/ut/FileStream.h
@@ -6,6 +6,13 @@
 namespace nw4hbm {
 namespace ut {
 
+#if defined(MSCHARGED_NATIVE)
+#pragma push_macro("SEEK_CUR")
+#pragma push_macro("SEEK_END")
+#undef SEEK_CUR
+#undef SEEK_END
+#endif
+
 class FileStream : public IOStream {
 public:
     enum SeekOrigin {
@@ -49,6 +56,11 @@ protected:
     }; // size = 0x08
 }; // size = 0x14
 
+#if defined(MSCHARGED_NATIVE)
+#pragma pop_macro("SEEK_END")
+#pragma pop_macro("SEEK_CUR")
+#endif
+
 } // namespace ut
 } // namespace nw4hbm
 
diff --git a/libs/RVL_SDK/include/revolution/hbm/nw4hbm/ut/inlines.h b/libs/RVL_SDK/include/revolution/hbm/nw4hbm/ut/inlines.h
index 886cc47..421a3bf 100644
--- a/libs/RVL_SDK/include/revolution/hbm/nw4hbm/ut/inlines.h
+++ b/libs/RVL_SDK/include/revolution/hbm/nw4hbm/ut/inlines.h
@@ -64,16 +64,26 @@ inline int ComparePtr(const void* pPtr1, const void* pPtr2) {
 template <typename T> inline T RoundUp(T t, u32 alignment) { return (alignment + t - 1) & ~(alignment - 1); }
 
 template <typename T> inline void* RoundUp(T* pPtr, u32 alignment) {
+#if defined(MSCHARGED_NATIVE)
+    std::uintptr_t value = reinterpret_cast<std::uintptr_t>(pPtr);
+    std::uintptr_t rounded = (std::uintptr_t(alignment) + value - 1) & ~std::uintptr_t(alignment - 1);
+#else
     u32 value = reinterpret_cast<u32>(pPtr);
     u32 rounded = (alignment + value - 1) & ~(alignment - 1);
+#endif
     return reinterpret_cast<void*>(rounded);
 }
 
 template <typename T> inline T RoundDown(T t, u32 alignment) { return t & ~(alignment - 1); }
 
 template <typename T> inline void* RoundDown(T* pPtr, u32 alignment) {
+#if defined(MSCHARGED_NATIVE)
+    std::uintptr_t value = reinterpret_cast<std::uintptr_t>(pPtr);
+    std::uintptr_t rounded = value & ~std::uintptr_t(alignment - 1);
+#else
     u32 value = reinterpret_cast<u32>(pPtr);
     u32 rounded = value & ~(alignment - 1);
+#endif
     return reinterpret_cast<void*>(rounded);
 }
 
diff --git a/src/RVL_SDK/hbm/nw4hbm/snd/snd_Channel.cpp b/src/RVL_SDK/hbm/nw4hbm/snd/snd_Channel.cpp
index 763cfcc..f7eda8e 100644
--- a/src/RVL_SDK/hbm/nw4hbm/snd/snd_Channel.cpp
+++ b/src/RVL_SDK/hbm/nw4hbm/snd/snd_Channel.cpp
@@ -1,3 +1,6 @@
+#if defined(MSCHARGED_NATIVE)
+#include "revolution/hbm/nw4hbm/ut/Lock.h"
+#endif
 #include "revolution/hbm/nw4hbm/snd/Channel.h"
 
 #include "revolution/ax.h"
diff --git a/src/RVL_SDK/hbm/nw4hbm/snd/snd_DisposeCallbackManager.cpp b/src/RVL_SDK/hbm/nw4hbm/snd/snd_DisposeCallbackManager.cpp
index 8d9d9ed..e928a61 100644
--- a/src/RVL_SDK/hbm/nw4hbm/snd/snd_DisposeCallbackManager.cpp
+++ b/src/RVL_SDK/hbm/nw4hbm/snd/snd_DisposeCallbackManager.cpp
@@ -1,6 +1,11 @@
 #include "revolution/hbm/nw4hbm/snd/DisposeCallbackManager.h"
 
+#if defined(MSCHARGED_NATIVE)
+#include "revolution/hbm/nw4hbm/ut/Lock.h"
+#include "revolution/hbm/HBMAssert.hpp"
+#else
 #include "revolution/hbm/nw4hbm/ut.h"
+#endif
 
 namespace nw4hbm {
 namespace snd {
diff --git a/src/RVL_SDK/hbm/nw4hbm/snd/snd_EnvGenerator.cpp b/src/RVL_SDK/hbm/nw4hbm/snd/snd_EnvGenerator.cpp
index 75ac319..590520c 100644
--- a/src/RVL_SDK/hbm/nw4hbm/snd/snd_EnvGenerator.cpp
+++ b/src/RVL_SDK/hbm/nw4hbm/snd/snd_EnvGenerator.cpp
@@ -1,4 +1,10 @@
+#if defined(MSCHARGED_NATIVE)
+#include "revolution/hbm/nw4hbm/snd/EnvGenerator.h"
+#include "revolution/hbm/nw4hbm/snd/Util.h"
+#include "revolution/hbm/HBMAssert.hpp"
+#else
 #include "revolution/hbm/nw4hbm/snd.h"
+#endif
 
 namespace nw4hbm {
 namespace snd {
diff --git a/src/RVL_SDK/hbm/nw4hbm/snd/snd_RemoteSpeaker.cpp b/src/RVL_SDK/hbm/nw4hbm/snd/snd_RemoteSpeaker.cpp
index 6ffc941..57a8e2d 100644
--- a/src/RVL_SDK/hbm/nw4hbm/snd/snd_RemoteSpeaker.cpp
+++ b/src/RVL_SDK/hbm/nw4hbm/snd/snd_RemoteSpeaker.cpp
@@ -1,4 +1,9 @@
+#if defined(MSCHARGED_NATIVE)
+#include "revolution/hbm/nw4hbm/ut/Lock.h"
+#include "revolution/hbm/HBMAssert.hpp"
+#else
 #include "revolution/hbm/nw4hbm/ut.h"
+#endif
 #include "revolution/hbm/nw4hbm/snd/RemoteSpeaker.h"
 
 #include "revolution/os.h"
diff --git a/src/RVL_SDK/hbm/nw4hbm/snd/snd_SeqPlayer.cpp b/src/RVL_SDK/hbm/nw4hbm/snd/snd_SeqPlayer.cpp
index abc0b3c..9bf2afc 100644
--- a/src/RVL_SDK/hbm/nw4hbm/snd/snd_SeqPlayer.cpp
+++ b/src/RVL_SDK/hbm/nw4hbm/snd/snd_SeqPlayer.cpp
@@ -4,7 +4,12 @@
 #include "revolution/hbm/nw4hbm/snd/NoteOnCallback.h"
 #include "revolution/hbm/nw4hbm/snd/SeqTrack.h"
 #include "revolution/hbm/nw4hbm/snd/SeqTrackAllocator.h"
+#if defined(MSCHARGED_NATIVE)
+#include "revolution/hbm/nw4hbm/ut/Lock.h"
+#include "revolution/hbm/HBMAssert.hpp"
+#else
 #include "revolution/hbm/nw4hbm/ut.h"
+#endif
 
 namespace nw4hbm {
 namespace snd {
diff --git a/src/RVL_SDK/hbm/nw4hbm/snd/snd_SeqSound.cpp b/src/RVL_SDK/hbm/nw4hbm/snd/snd_SeqSound.cpp
index 54c3f5e..c675333 100644
--- a/src/RVL_SDK/hbm/nw4hbm/snd/snd_SeqSound.cpp
+++ b/src/RVL_SDK/hbm/nw4hbm/snd/snd_SeqSound.cpp
@@ -1,6 +1,19 @@
 #include "decomp.h"
+#if defined(MSCHARGED_NATIVE)
+#include "revolution/hbm/nw4hbm/snd/BasicSound.h"
+#include "revolution/hbm/nw4hbm/snd/SeqPlayer.h"
+#include "revolution/hbm/nw4hbm/snd/SeqSound.h"
+#include "revolution/hbm/nw4hbm/snd/SeqSoundHandle.h"
+#include "revolution/hbm/nw4hbm/snd/SoundInstanceManager.h"
+#else
 #include "revolution/hbm/nw4hbm/snd.h"
+#endif
+#if defined(MSCHARGED_NATIVE)
+#include "revolution/hbm/nw4hbm/ut/Lock.h"
+#include "revolution/hbm/HBMAssert.hpp"
+#else
 #include "revolution/hbm/nw4hbm/ut.h"
+#endif
 
 namespace nw4hbm {
 namespace snd {
diff --git a/src/RVL_SDK/hbm/nw4hbm/snd/snd_SeqTrack.cpp b/src/RVL_SDK/hbm/nw4hbm/snd/snd_SeqTrack.cpp
index 3f2c116..e56f679 100644
--- a/src/RVL_SDK/hbm/nw4hbm/snd/snd_SeqTrack.cpp
+++ b/src/RVL_SDK/hbm/nw4hbm/snd/snd_SeqTrack.cpp
@@ -1,6 +1,17 @@
 #include "decomp.h"
+#if defined(MSCHARGED_NATIVE)
+#include "revolution/hbm/nw4hbm/snd/SeqTrack.h"
+#include "revolution/hbm/nw4hbm/snd/SeqPlayer.h"
+#include "revolution/hbm/nw4hbm/snd/NoteOnCallback.h"
+#else
 #include "revolution/hbm/nw4hbm/snd.h"
+#endif
+#if defined(MSCHARGED_NATIVE)
+#include "revolution/hbm/nw4hbm/ut/Lock.h"
+#include "revolution/hbm/HBMAssert.hpp"
+#else
 #include "revolution/hbm/nw4hbm/ut.h"
+#endif
 
 namespace nw4hbm {
 namespace snd {
diff --git a/src/RVL_SDK/hbm/nw4hbm/snd/snd_SoundArchivePlayer.cpp b/src/RVL_SDK/hbm/nw4hbm/snd/snd_SoundArchivePlayer.cpp
index dc08b12..2f66291 100644
--- a/src/RVL_SDK/hbm/nw4hbm/snd/snd_SoundArchivePlayer.cpp
+++ b/src/RVL_SDK/hbm/nw4hbm/snd/snd_SoundArchivePlayer.cpp
@@ -1,3 +1,7 @@
+#if defined(MSCHARGED_NATIVE)
+#include "revolution/dvd/dvdqueue.h"
+#include "revolution/hbm/nw4hbm/ut/DvdFileStream.h"
+#endif
 
 
 #include "revolution/hbm/nw4hbm/snd/SoundArchivePlayer.h"
diff --git a/src/RVL_SDK/hbm/nw4hbm/snd/snd_StrmChannel.cpp b/src/RVL_SDK/hbm/nw4hbm/snd/snd_StrmChannel.cpp
index b1a26e6..18243bb 100644
--- a/src/RVL_SDK/hbm/nw4hbm/snd/snd_StrmChannel.cpp
+++ b/src/RVL_SDK/hbm/nw4hbm/snd/snd_StrmChannel.cpp
@@ -1,5 +1,14 @@
+#if defined(MSCHARGED_NATIVE)
+#include "revolution/hbm/nw4hbm/snd/StrmChannel.h"
+#else
 #include "revolution/hbm/nw4hbm/snd.h"
+#endif
+#if defined(MSCHARGED_NATIVE)
+#include "revolution/hbm/nw4hbm/ut/Lock.h"
+#include "revolution/hbm/HBMAssert.hpp"
+#else
 #include "revolution/hbm/nw4hbm/ut.h"
+#endif
 
 #include <string.h>
 
diff --git a/src/RVL_SDK/hbm/nw4hbm/snd/snd_StrmPlayer.cpp b/src/RVL_SDK/hbm/nw4hbm/snd/snd_StrmPlayer.cpp
index 8caf49d..9bbbceb 100644
--- a/src/RVL_SDK/hbm/nw4hbm/snd/snd_StrmPlayer.cpp
+++ b/src/RVL_SDK/hbm/nw4hbm/snd/snd_StrmPlayer.cpp
@@ -3,7 +3,12 @@
 #include "revolution/hbm/nw4hbm/snd/WaveFile.h"
 
 #include "revolution/hbm/nw4hbm/snd/AxManager.h"
+#if defined(MSCHARGED_NATIVE)
+#include "revolution/hbm/nw4hbm/ut/Lock.h"
+#include "revolution/hbm/HBMAssert.hpp"
+#else
 #include "revolution/hbm/nw4hbm/ut.h"
+#endif
 
 #include <string.h>
 #include <limits.h>
diff --git a/src/RVL_SDK/hbm/nw4hbm/snd/snd_Util.cpp b/src/RVL_SDK/hbm/nw4hbm/snd/snd_Util.cpp
index 37eebc5..4441ba7 100644
--- a/src/RVL_SDK/hbm/nw4hbm/snd/snd_Util.cpp
+++ b/src/RVL_SDK/hbm/nw4hbm/snd/snd_Util.cpp
@@ -1,7 +1,14 @@
 #include <math.h>
+#if !defined(MSCHARGED_NATIVE)
 #include "revolution/hbm/nw4hbm/math.h"
+#endif
 #include "revolution/hbm/nw4hbm/snd/Util.h"
+#if defined(MSCHARGED_NATIVE)
+#include "revolution/hbm/nw4hbm/ut/Lock.h"
+#include "revolution/hbm/HBMAssert.hpp"
+#else
 #include "revolution/hbm/nw4hbm/ut.h"
+#endif
 
 namespace nw4hbm {
 namespace snd {
diff --git a/src/RVL_SDK/hbm/nw4hbm/snd/snd_WsdPlayer.cpp b/src/RVL_SDK/hbm/nw4hbm/snd/snd_WsdPlayer.cpp
index 9915c0c..8264ade 100644
--- a/src/RVL_SDK/hbm/nw4hbm/snd/snd_WsdPlayer.cpp
+++ b/src/RVL_SDK/hbm/nw4hbm/snd/snd_WsdPlayer.cpp
@@ -6,7 +6,12 @@
 
 #include "revolution/hbm/nw4hbm/snd/DisposeCallbackManager.h"
 
+#if defined(MSCHARGED_NATIVE)
+#include "revolution/hbm/nw4hbm/ut/Lock.h"
+#include "revolution/hbm/HBMAssert.hpp"
+#else
 #include "revolution/hbm/nw4hbm/ut.h"
+#endif
 
 #include "decomp.h"
 
diff --git a/src/RVL_SDK/hbm/nw4hbm/snd/snd_WsdTrack.cpp b/src/RVL_SDK/hbm/nw4hbm/snd/snd_WsdTrack.cpp
index 56ab174..cfc6791 100644
--- a/src/RVL_SDK/hbm/nw4hbm/snd/snd_WsdTrack.cpp
+++ b/src/RVL_SDK/hbm/nw4hbm/snd/snd_WsdTrack.cpp
@@ -3,7 +3,12 @@
 
 #include "revolution/hbm/nw4hbm/snd/AxManager.h"
 
+#if defined(MSCHARGED_NATIVE)
+#include "revolution/hbm/nw4hbm/ut/Lock.h"
+#include "revolution/hbm/HBMAssert.hpp"
+#else
 #include "revolution/hbm/nw4hbm/ut.h"
+#endif
 
 namespace nw4hbm {
 namespace snd {
